### PR TITLE
When we pull in a cacheline, we should use it too.

### PR DESCRIPTION
**Pull request by baldersheim**
There is possibly wasting 7/8 of it and very likely suffer a cache miss.

@havardpe PR

### CODE REVIEW

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp [hunk 1]
```diff
@@ -101,6 +104,82 @@ void verifyPopulationCount(const IAccelrated & accel)
...
+    }
+    for (size_t offset = 0; offset < 8; offset++) {
+        for (size_t i = 1; i < VESPA_NELEMS(vectors); i++) {
+            std::vector<uint64_t> expected = vectors[0];
+            for (size_t j = 1; j < i; j++) {
+                simpleOrWith(expected, vectors[j]);
+            }
+            std::vector<std::pair<const uint64_t *, bool>> vRefs;
+            for (size_t j(0); j < i; j++) {
+                vRefs.emplace_back(&vectors[j][0], false);
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp:143]: should also test with inverted vectors

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp [hunk 2]
```diff
@@ -101,6 +104,82 @@ void verifyPopulationCount(const IAccelrated & accel)
...
+    }
+    for (size_t offset = 0; offset < 8; offset++) {
+        for (size_t i = 1; i < VESPA_NELEMS(vectors); i++) {
+            std::vector<uint64_t> expected = vectors[0];
+            for (size_t j = 1; j < i; j++) {
+                simpleAndWith(expected, vectors[j]);
+            }
+            std::vector<std::pair<const uint64_t *, bool>> vRefs;
+            for (size_t j(0); j < i; j++) {
+                vRefs.emplace_back(&vectors[j][0], false);
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp:170]: should also test with inverted vectors

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp [hunk 3]
```diff
@@ -101,6 +104,82 @@ void verifyPopulationCount(const IAccelrated & accel)
...
+            }
+            uint64_t dest[8] __attribute((aligned(64)));
+            accel.and64(offset, vRefs, dest);
+            int diff = memcmp(&expected[offset], dest, sizeof(dest));
+            if (diff != 0) {
+                fprintf(stderr, "Accelrator is not failing and64\n");
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp:176]: misleading error message

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp [hunk 4]
```diff
@@ -101,6 +104,82 @@ void verifyPopulationCount(const IAccelrated & accel)
...
+            }
+            std::vector<std::pair<const uint64_t *, bool>> vRefs;
+            for (size_t j(0); j < i; j++) {
+                vRefs.emplace_back(&vectors[j][0], false);
+            }
+            uint64_t dest[8] __attribute((aligned(64)));
+            accel.or64(offset, vRefs, dest);
+            int diff = memcmp(&expected[offset], dest, sizeof(dest));
+            if (diff != 0) {
+                fprintf(stderr, "Accelrator is not failing and64\n");
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp:149]: even more misleading error message

File: searchlib/src/vespa/searchlib/queryeval/multibitvectoriterator.cpp [hunk 5]
```diff
@@ -1,61 +1,90 @@
...
 protected:
     void updateLastValue(uint32_t docId);
     void strictSeek(uint32_t docId);
 private:
     void doSeek(uint32_t docId) override;
     Trinary is_strict() const override { return Trinary::False; }
     bool acceptExtraFilter() const override { return Update::isAnd(); }
-    Update                  _update;
+    Update              _update;
+    Word                _lastWords[8] __attribute__((aligned(32)));
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/multibitvectoriterator.cpp:40]: should add a static assert somewhere that this value is 64 bytes

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.h [hunk 6]
```diff
@@ -29,6 +30,9 @@ class IAccelrated
     virtual size_t populationCount(const uint64_t *a, size_t sz) const = 0;
     virtual double squaredEuclideanDistance(const float * a, const float * b, size_t sz) const = 0;
     virtual double squaredEuclideanDistance(const double * a, const double * b, size_t sz) const = 0;
+    // And 64 bytes from multiple sources
+    virtual void and64(size_t offset, const std::vector<std::pair<const uint64_t *, bool>> &src, uint64_t *dest) const = 0;
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.h:34]: these functions are a bit confusing. The length of the operation is specified in bytes while the offset is specified in multiples of uint64_t. It also binds the Word abstraction to be uint64_t, making that abstraction less useful. Also; at glace, the 64 in the function name seems to reflect the fact that we are using 64-bit values, leaving the size of the operation a mystery...

consider using byte offset and passing the memory as void*

File: searchlib/src/vespa/searchlib/queryeval/multibitvectoriterator.cpp (hunk 5 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/multibitvectoriterator.cpp:40]: Fixed

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp (hunk 4 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp:149]: Fixed

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp (hunk 3 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp:176]: Fixed

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp (hunk 2 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp]: Fixed

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp (hunk 1 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp]: Fixed

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.h (hunk 6 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.h:34]: Agree, fixed

**Pull request merged**